cpp/ql/lib/experimental/semmle/code/cpp/dataflow/ProductFlow.qll

```diff
@@ -0,0 +1,119 @@
+
+    predicate hasFlowPath(
+      DataFlow::PathNode source1, DataFlow2::PathNode source2, DataFlow::PathNode sink1,
+      DataFlow2::PathNode sink2
+    ) {
+      reachable(this, source1, source2, sink1, sink2)
+    }
+  }
+
+  private import Internal
+
+  module Internal {
+    class Conf1 extends DataFlow::Configuration {
+      Conf1() { this = "Conf1" }
+
+      override predicate isSource(DataFlow::Node source) {
+        exists(Configuration conf | conf.isSourcePair(source, _))
+      }
+
+      override predicate isSink(DataFlow::Node sink) {
+        exists(Configuration conf | conf.isSinkPair(sink, _))
+      }
+    }
+
+    class Conf2 extends DataFlow2::Configuration {
+      Conf2() { this = "Conf2" }
+
+      override predicate isSource(DataFlow::Node source) {
+        exists(Configuration conf, DataFlow::Node source1 |
+          conf.isSourcePair(source1, source) and
+          any(Conf1 c).hasFlow(source1, _)
+        )
+      }
+
+      override predicate isSink(DataFlow::Node sink) {
+        exists(Configuration conf, DataFlow::Node sink1 |
+          conf.isSinkPair(sink1, sink) and any(Conf1 c).hasFlow(_, sink1)
+        )
+      }
+    }
+  }
+
+
+  private predicate reachableInterprocEntry(
+    Configuration conf, DataFlow::PathNode source1, DataFlow2::PathNode source2,
+    DataFlow::PathNode node1, DataFlow2::PathNode node2
+  ) {
+    conf.isSourcePair(node1.getNode(), node2.getNode()) and
+    node1 = source1 and
+    node2 = source2
+    or
+    exists(
+      DataFlow::PathNode midEntry1, DataFlow2::PathNode midEntry2, DataFlow::PathNode midExit1,
+      DataFlow2::PathNode midExit2
+    |
+      reachableInterprocEntry(conf, source1, source2, midEntry1, midEntry2) and
+      interprocEdgePair(midExit1, midExit2, node1, node2) and
+      localPathStep1*(midEntry1, midExit1) and
+      localPathStep2*(midEntry2, midExit2)
+    )
+  }
+
+  private predicate localPathStep1(DataFlow::PathNode pred, DataFlow::PathNode succ) {
+    DataFlow::PathGraph::edges(pred, succ) and
+    pred.getNode().getEnclosingCallable() = succ.getNode().getEnclosingCallable()
+  }
+
+  private predicate localPathStep2(DataFlow2::PathNode pred, DataFlow2::PathNode succ) {
+    DataFlow2::PathGraph::edges(pred, succ) and
+    pred.getNode().getEnclosingCallable() = succ.getNode().getEnclosingCallable()
+  }
+
+  private predicate interprocEdgePair(
+    DataFlow::PathNode pred1, DataFlow2::PathNode pred2, DataFlow::PathNode succ1,
+    DataFlow2::PathNode succ2
+  ) {
+    exists(Declaration predDecl, Declaration succDecl |
+      DataFlow::PathGraph::edges(pred1, succ1) and
+      DataFlow2::PathGraph::edges(pred2, succ2) and
+      predDecl != succDecl and
+      pred1.getNode().getEnclosingCallable() = predDecl and
+      pred2.getNode().getEnclosingCallable() = predDecl and
+      succ1.getNode().getEnclosingCallable() = succDecl and
+      succ2.getNode().getEnclosingCallable() = succDecl
+    )
+  }
```

Comment:
Yes, this predicate is exactly what I had in mind. But as written, its best join-order isn't pipelineable. So it'll very likely need to be factored like this:
```suggestion
  pragma[nomagic]
  private predicate interprocEdge1(Declaration predDecl, Declaration succDecl, DataFlow::PathNode pred1, DataFlow::PathNode succ1) {
    DataFlow::PathGraph::edges(pred1, succ1) and
    predDecl != succDecl and
    pred1.getNode().getEnclosingCallable() = predDecl and
    succ1.getNode().getEnclosingCallable() = succDecl
  }

  pragma[nomagic]
  private predicate interprocEdge2(Declaration predDecl, Declaration succDecl, DataFlow2::PathNode pred2, DataFlow2::PathNode succ2) {
    DataFlow2::PathGraph::edges(pred2, succ2) and
    predDecl != succDecl and
    pred2.getNode().getEnclosingCallable() = predDecl and
    succ2.getNode().getEnclosingCallable() = succDecl
  }

  private predicate interprocEdgePair(
    DataFlow::PathNode pred1, DataFlow2::PathNode pred2, DataFlow::PathNode succ1,
    DataFlow2::PathNode succ2
  ) {
    exists(Declaration predDecl, Declaration succDecl |
      interprocEdge1(predDecl, succDecl, pred1, succ1) and
      interprocEdge2(predDecl, succDecl, pred2, succ2)
    )
  }
```

Reply:
Fixed in e0a5d18d7d9a0e0807a3a1191b94d73ae6af7bd3.